src/broadcaster/queue_broadcaster.rs

```diff
@@ -0,0 +1,273 @@
+use std::iter;
+
+use cosmrs::{Any, Gas};
+use error_stack::{self, Report, ResultExt};
+use thiserror::Error;
+use tokio::{select, sync::mpsc};
+
+use super::{clients::BroadcastClient, msg_queue::MsgQueue, Broadcaster};
+
+type Result<T = ()> = error_stack::Result<T, Error>;
+
+#[derive(Error, Debug)]
+pub enum Error {
+    #[error("failed estimating fee for message")]
+    EstimateFee,
+    #[error("failed broadcasting messages in queue")]
+    Broadcast,
+    #[error("failed closing client")]
+    Close,
+}
+
+pub struct QueuedBroadcasterDriver {
+    broadcast_tx: mpsc::Sender<()>,
+}
+
+impl QueuedBroadcasterDriver {
+    pub async fn force_broadcast(&self) -> Result {
+        self.broadcast_tx
+            .send(())
+            .await
+            .map_err(|_| Report::new(Error::Broadcast))
+    }
+}
+
+pub struct QueuedBroadcaster<T>
+where
+    T: BroadcastClient,
+{
+    broadcaster: Broadcaster<T>,
+    queue: MsgQueue,
+    gas_limit: Gas,
+    channel: (mpsc::Sender<Any>, mpsc::Receiver<Any>),
+    broadcast_rx: mpsc::Receiver<()>,
+}
+
+impl<T> QueuedBroadcaster<T>
+where
+    T: BroadcastClient,
+{
+    pub fn new(broadcaster: Broadcaster<T>, gas_limit: Gas, capacity: usize) -> (Self, QueuedBroadcasterDriver) {
+        let (broadcast_tx, broadcast_rx) = mpsc::channel(1);
+
+        (
+            Self {
+                broadcaster,
+                queue: MsgQueue::default(),
+                gas_limit,
+                channel: mpsc::channel(capacity),
+                broadcast_rx,
+            },
+            QueuedBroadcasterDriver { broadcast_tx },
+        )
+    }
+
+    pub async fn run(mut self) -> Result {
+        let (tx, mut rx) = self.channel;
+        drop(tx);
+
+        let mut queue = self.queue;
+        let mut broadcaster = self.broadcaster;
+
+        loop {
+            select! {
+              msg = rx.recv() => match msg {
+                None => break,
+                Some(msg) => {
+                  let fee = broadcaster.estimate_fee(iter::once(msg.clone())).await.change_context(Error::EstimateFee)?;
+
+                  if fee.gas_limit + queue.gas_cost() > self.gas_limit {
+                    broadcast_all(&mut queue, &mut broadcaster).await?;
+                  }
+
+                  queue.push(msg, fee.gas_limit);
+                }
+              },
+              _ = self.broadcast_rx.recv() => broadcast_all(&mut queue, &mut broadcaster).await?,
+            }
+        }
+
+        Ok(())
+    }
+
+    pub fn tx_sender(&self) -> mpsc::Sender<Any> {
```

Comment:
```suggestion
    pub fn endpoint(&self) -> mpsc::Sender<Any> {
```
I don't really like endpoint either, it's really hard to find a good name for this function 😆

Reply:
```suggestion
    pub fn client(&self) -> BroadcastClient {
```
and have a thin wrapper around the channel, so you could call `client.Broadcast(Any)`?